### PR TITLE
Add table aws_kinesis_firehose_delivery_stream. closes #195

### LINKED ISSUE #195: Add table aws_kinesis_firehose_delivery_stream
**References**
https://pkg.go.dev/github.com/aws/aws-sdk-go/service/firehose

### PR DESCRIPTION
**Pull request by sayan133**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_kinesis_firehose_delivery_stream []

PRETEST: tests/aws_kinesis_firehose_delivery_stream

TEST: tests/aws_kinesis_firehose_delivery_stream
Running terraform
aws_s3_bucket.firehose_bucket: Creating...
aws_iam_role.firehose_role: Creating...
aws_iam_role.firehose_role: Creation complete after 6s [id=firehose_test_role]
aws_s3_bucket.firehose_bucket: Still creating... [10s elapsed]
aws_s3_bucket.firehose_bucket: Still creating... [20s elapsed]
aws_s3_bucket.firehose_bucket: Creation complete after 20s [id=tf-firehose-bucket]
aws_kinesis_firehose_delivery_stream.named_test_resource: Creating...
aws_kinesis_firehose_delivery_stream.named_test_resource: Still creating... [10s elapsed]
aws_kinesis_firehose_delivery_stream.named_test_resource: Still creating... [20s elapsed]
aws_kinesis_firehose_delivery_stream.named_test_resource: Still creating... [30s elapsed]
aws_kinesis_firehose_delivery_stream.named_test_resource: Creation complete after 40s [id=arn:aws:firehose:us-east-1:388460667113:deliverystream/turbottest28655]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 3 added, 0 changed, 0 destroyed.

Outputs:

account_id = "388460667113"
aws_partition = "aws"
aws_region = "us-east-1"
bucket_arn = "arn:aws:s3:::tf-firehose-bucket"
resource_aka = "arn:aws:firehose:us-east-1:388460667113:deliverystream/turbottest28655"
resource_name = "turbottest28655"
role_arn = "arn:aws:iam::388460667113:role/firehose_test_role"

Running SQL query: test-get-query.sql
[
  {
    "account_id": "388460667113",
    "delivery_stream_arn": "arn:aws:firehose:us-east-1:388460667113:deliverystream/turbottest28655",
    "delivery_stream_encryption_configuration": {
      "FailureDescription": null,
      "KeyARN": null,
      "KeyType": null,
      "Status": "DISABLED"
    },
    "delivery_stream_name": "turbottest28655",
    "delivery_stream_status": "ACTIVE",
    "delivery_stream_type": "DirectPut",
    "destinations": [
      {
        "DestinationId": "destinationId-000000000001",
        "ElasticsearchDestinationDescription": null,
        "ExtendedS3DestinationDescription": {
          "BucketARN": "arn:aws:s3:::tf-firehose-bucket",
          "BufferingHints": {
            "IntervalInSeconds": 300,
            "SizeInMBs": 5
          },
          "CloudWatchLoggingOptions": {
            "Enabled": false,
            "LogGroupName": null,
            "LogStreamName": null
          },
          "CompressionFormat": "UNCOMPRESSED",
          "DataFormatConversionConfiguration": null,
          "EncryptionConfiguration": {
            "KMSEncryptionConfig": null,
            "NoEncryptionConfig": "NoEncryption"
          },
          "ErrorOutputPrefix": null,
          "Prefix": "",
          "ProcessingConfiguration": null,
          "RoleARN": "arn:aws:iam::388460667113:role/firehose_test_role",
          "S3BackupDescription": null,
          "S3BackupMode": "Disabled"
        },
        "HttpEndpointDestinationDescription": null,
        "RedshiftDestinationDescription": null,
        "S3DestinationDescription": {
          "BucketARN": "arn:aws:s3:::tf-firehose-bucket",
          "BufferingHints": {
            "IntervalInSeconds": 300,
            "SizeInMBs": 5
          },
          "CloudWatchLoggingOptions": {
            "Enabled": false,
            "LogGroupName": null,
            "LogStreamName": null
          },
          "CompressionFormat": "UNCOMPRESSED",
          "EncryptionConfiguration": {
            "KMSEncryptionConfig": null,
            "NoEncryptionConfig": "NoEncryption"
          },
          "ErrorOutputPrefix": null,
          "Prefix": "",
          "RoleARN": "arn:aws:iam::388460667113:role/firehose_test_role"
        },
        "SplunkDestinationDescription": null
      }
    ],
    "failure_description": "<null>",
    "has_more_destinations": false,
    "last_update_timestamp": "<null>",
    "partition": "aws",
    "region": "us-east-1",
    "source": "<null>",
    "tags_src": [
      {
        "Key": "Foo",
        "Value": "Bar"
      },
      {
        "Key": "Name",
        "Value": "turbottest28655"
      }
    ],
    "version_id": "1"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "delivery_stream_arn": "arn:aws:firehose:us-east-1:388460667113:deliverystream/turbottest28655",
    "delivery_stream_name": "turbottest28655",
    "delivery_stream_status": "ACTIVE",
    "version_id": "1"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:firehose:us-east-1:388460667113:deliverystream/turbottest28655"
    ],
    "tags": {
      "Foo": "Bar",
      "Name": "turbottest28655"
    },
    "title": "turbottest28655"
  }
]
✔ PASSED

POSTTEST: tests/aws_kinesis_firehose_delivery_stream

TEARDOWN: tests/aws_kinesis_firehose_delivery_stream

SUMMARY:

1/1 passed.

```
</details>

# Example query results
<details>
  <summary>Results</summary>
  

### Basic info
```
select
  delivery_stream_name,
  arn,
  create_timestamp,
  delivery_stream_type
from
  aws_kinesis_firehose_delivery_stream;
```
```
+----------------------+-----------------------------------------------------------------------+---------------------+-----------------------+
| delivery_stream_name | arn                                                   | create_timestamp    | delivery_stream_type  |
+----------------------+-----------------------------------------------------------------------+---------------------+-----------------------+
| test                 | arn:aws:firehose:us-east-2:388460667113:deliverystream/test           | 2021-03-18 08:41:48 | DirectPut             |
| encrypted-test       | arn:aws:firehose:us-east-2:388460667113:deliverystream/encrypted-test | 2021-03-18 14:12:19 | KinesisStreamAsSource |
+----------------------+-----------------------------------------------------------------------+---------------------+-----------------------+
```

### List all firehose delivery streams which are not in ACTIVE state
```
select
  delivery_stream_name,
  arn,
  delivery_stream_status,
  create_timestamp,
  delivery_stream_type
from
  aws_kinesis_firehose_delivery_stream
where
  delivery_stream_status != 'ACTIVE';
```
```
+----------------------+---------------------+------------------------+------------------+----------------------+
| delivery_stream_name | arn | delivery_stream_status | create_timestamp | delivery_stream_type |
+----------------------+---------------------+------------------------+------------------+----------------------+
+----------------------+---------------------+------------------------+------------------+----------------------+
```

### List all firehose delivery streams for a specific delivery stream type
```
select
  delivery_stream_name,
  arn,
  delivery_stream_status,
  create_timestamp,
  delivery_stream_type
from
  aws_kinesis_firehose_delivery_stream
where
  delivery_stream_type = 'DirectPut';
```
```
+----------------------+-------------------------------------------------------------+------------------------+---------------------+----------------------+
| delivery_stream_name | arn                                         | delivery_stream_status | create_timestamp    | delivery_stream_type |
+----------------------+-------------------------------------------------------------+------------------------+---------------------+----------------------+
| test                 | arn:aws:firehose:us-east-2:388460667113:deliverystream/test | ACTIVE                 | 2021-03-18 08:41:48 | DirectPut            |
+----------------------+-------------------------------------------------------------+------------------------+---------------------+----------------------+
```

### List all firehose delivery streams which updated in last 7 days
```
select
  delivery_stream_name,
  arn,
  delivery_stream_status,
  create_timestamp,
  last_update_timestamp,
  delivery_stream_type
from
  aws_kinesis_firehose_delivery_stream
where
  last_update_timestamp > (current_date - interval '7' day);
```
```
+----------------------+-------------------------------------------------------------+------------------------+---------------------+-----------------------+----------------------+
| delivery_stream_name | arn                                         | delivery_stream_status | create_timestamp    | last_update_timestamp | delivery_stream_type |
+----------------------+-------------------------------------------------------------+------------------------+---------------------+-----------------------+----------------------+
| test                 | arn:aws:firehose:us-east-2:388460667113:deliverystream/test | ACTIVE                 | 2021-03-18 08:41:48 | 2021-03-18 14:13:50   | DirectPut            |
+----------------------+-------------------------------------------------------------+------------------------+---------------------+-----------------------+----------------------+
```

### List all firehose delivery streams which are not encrypted
```
select
  delivery_stream_name,
  arn,
  delivery_stream_status,
  create_timestamp,
  delivery_stream_type,
  delivery_stream_encryption_configuration ->> 'Status' as encryption_status
from
  aws_kinesis_firehose_delivery_stream
where
  delivery_stream_encryption_configuration ->> 'Status' = 'DISABLED';
```
```
+----------------------+-------------------------------------------------------------+------------------------+---------------------+----------------------+-------------------+
| delivery_stream_name | arn                                                         | delivery_stream_status | create_timestamp    | delivery_stream_type | encryption_status |
+----------------------+-------------------------------------------------------------+------------------------+---------------------+----------------------+-------------------+
| test                 | arn:aws:firehose:us-east-1:388460667113:deliverystream/test | ACTIVE                 | 2021-03-30 11:22:10 | DirectPut            | DISABLED          |
+----------------------+-------------------------------------------------------------+------------------------+---------------------+----------------------+-------------------+
```

### List all firehose delivery streams with failure description
```
select
  delivery_stream_name,
  arn,
  delivery_stream_status,
  delivery_stream_type,
  failure_description
from
  aws_kinesis_firehose_delivery_stream
where
  failure_description is not null;
```
```
+----------------------+---------------------+------------------------+----------------------+---------------------+
| delivery_stream_name | arn | delivery_stream_status | delivery_stream_type | failure_description |
+----------------------+---------------------+------------------------+----------------------+---------------------+
+----------------------+---------------------+------------------------+----------------------+---------------------+
```
</details>

### CODE REVIEW

File: aws/table_aws_kinesis_firehose_delivery_stream.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+		},
+		GetMatrixItem: BuildRegionList,
+		Columns: awsRegionalColumns([]*plugin.Column{
+			{
+				Name:        "delivery_stream_name",
+				Description: "The name of the delivery stream.",
+				Type:        pb.ColumnType_STRING,
+			},
+			{
+				Name:        "delivery_stream_arn",
```
Comment (Paulami30) [inline aws/table_aws_kinesis_firehose_delivery_stream.go:35]:
```suggestion
				Name:        "arn",
```

File: aws/table_aws_kinesis_firehose_delivery_stream.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+				Hydrate:     describeStream,
+				Transform:   transform.FromField("DeliveryStreamDescription.DeliveryStreamARN").Transform(arnToAkas),
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listStreams(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```
Comment (Paulami30) [inline aws/table_aws_kinesis_firehose_delivery_stream.go:147]:
```suggestion
func listFirehoseDeliveryStreams(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_aws_kinesis_firehose_delivery_stream.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+			break
+		}
+	}
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func describeStream(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Paulami30) [inline aws/table_aws_kinesis_firehose_delivery_stream.go:184]:
```suggestion
func describeFirehoseDeliveryStreams(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_aws_kinesis_firehose_delivery_stream.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+	data, err := svc.DescribeDeliveryStream(params)
+	if err != nil {
+		logger.Debug("describeDeliveryStream__", "ERROR", err)
+		return nil, err
+	}
+	return data, nil
+}
+
+// API call for fetching tag list
+func getAwsKinesisFirehoseDeliveryStreamTags(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Paulami30) [inline aws/table_aws_kinesis_firehose_delivery_stream.go:223]:
```suggestion
func listFirehoseDeliveryStreamTags(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
+# Table: aws_kinesis_firehose_delivery_stream
+
+The AWS Kinesis Firehose Delivery Stream resource delivers real-time streaming data to an Amazon Simple Storage Service (Amazon S3), Amazon Redshift, or Amazon Elasticsearch Service (Amazon ES) destination.
```
Comment (Paulami30) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:3]:
```suggestion
The AWS Kinesis Firehose Delivery Stream resource delivers real-time streaming data to an Amazon Simple Storage Service (Amazon S3), Amazon Redshift or Amazon Elasticsearch Service (Amazon ES) destination.
```

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  delivery_stream_name,
+  delivery_stream_arn,
+  create_timestamp,
+  delivery_stream_type
+from
+  aws_kinesis_firehose_delivery_stream;
+```
+
+
+### List firehose delivery streams which are not in ACTIVE state
```
Comment (Paulami30) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:20]:
```suggestion
### List firehose delivery streams which are not active
```

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+The AWS Kinesis Firehose Delivery Stream resource delivers real-time streaming data to an Amazon Simple Storage Service (Amazon S3), Amazon Redshift, or Amazon Elasticsearch Service (Amazon ES) destination.
+
+## Examples
+
+### Basic info
+
+```sql
+select
+  delivery_stream_name,
+  delivery_stream_arn,
```
Comment (Paulami30) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:12]:
```suggestion
  arn,
```

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+  create_timestamp,
+  delivery_stream_type
+from
+  aws_kinesis_firehose_delivery_stream
+where
+  delivery_stream_type = 'DirectPut';
+```
+
+
+### List firehose delivery streams which updated in last 7 days
```
Comment (Paulami30) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:52]:
```suggestion
### List firehose delivery streams which are not updated in last 7 days
```

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+  delivery_stream_arn,
+  delivery_stream_status,
+  create_timestamp,
+  delivery_stream_type,
+  delivery_stream_encryption_configuration ->> 'Status' as encryption_status
+from
+  aws_kinesis_firehose_delivery_stream
+where
+  delivery_stream_encryption_configuration ->> 'Status' = 'ENABLED';
+```
```
Comment (Paulami30) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:83]: Make the above query as your 3rd query and bring the current 3rd query here

File: aws/table_aws_kinesis_firehose_delivery_stream.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+				Hydrate:     describeStream,
+				Transform:   transform.FromField("DeliveryStreamDescription.Destinations"),
+			},
+			{
+				Name:        "failure_description",
+				Description: "Provides details in case one of the following operations fails due to an error related to KMS: CreateDeliveryStream, DeleteDeliveryStream, StartDeliveryStreamEncryption,StopDeliveryStreamEncryption.",
+				Type:        pb.ColumnType_JSON,
+				Hydrate:     describeStream,
+				Transform:   transform.FromField("DeliveryStreamDescription.FailureDescription"),
+			},
```
Comment (Paulami30) [inline aws/table_aws_kinesis_firehose_delivery_stream.go:110]: leaving the first 5 columns, can you please arrange the remaining columns in alphabetical order

**Review by Paulami30 — COMMENTED** [review]
please check the comments and resolve the merge conflicts

File: aws/service.go [hunk 11]
```diff
@@ -412,6 +412,26 @@ func EmrService(ctx context.Context, d *plugin.QueryData, region string) (*emr.E
 	return svc, nil
 }
 
+// Firehose returns the service connection for AWS Kinesis Firehose service
```
Comment (Paulami30) [inline aws/service.go:415]:
```suggestion
// FirehoseService returns the service connection for AWS Kinesis Firehose service
```

File: aws/table_aws_kinesis_firehose_delivery_stream.go [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+			break
+		}
+	}
+
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func describeFirehoseDeliveryStreams(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Paulami30) [inline aws/table_aws_kinesis_firehose_delivery_stream.go:184]:
```suggestion
func describeFirehoseDeliveryStream(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  delivery_stream_name,
+  arn,
+  create_timestamp,
+  delivery_stream_type
+from
+  aws_kinesis_firehose_delivery_stream;
+```
+
+
+### List firehose delivery streams which are not in active
```
Comment (Paulami30) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:20]:
```suggestion
### List firehose delivery streams which are not active
```

**Comment by Paulami30** [comment]
LGTM

**Review by Paulami30 — APPROVED** [review]

File: docs/tables/aws_kinesis_firehose_delivery_stream.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+  delivery_stream_name,
+  arn,
+  delivery_stream_status,
+  create_timestamp,
+  delivery_stream_type,
+  delivery_stream_encryption_configuration ->> 'Status' as encryption_status
+from
+  aws_kinesis_firehose_delivery_stream
+where
+  delivery_stream_encryption_configuration ->> 'Status' = 'ENABLED';
```
Comment (rajmohanty17) [inline docs/tables/aws_kinesis_firehose_delivery_stream.md:49]: where
  delivery_stream_encryption_configuration ->> 'Status' = 'ENABLED'; 
OR != 'ENABLED' ?

**Review by rajmohanty17 — CHANGES_REQUESTED** [review]
Check the List firehose delivery streams which are not encrypted where clause.

**Review by rajmohanty17 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**